Scripting: Execute Painless API fails when communicating across nodes

**Elasticsearch version** (`bin/elasticsearch --version`): 6.5.1

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

Start two elasticsearch, one named `first`, one named `second`. The `first` one should be the where you connect kibana to, to make sure this works (just start it as the first one, so it will bind to port 9200)

```
 ./elasticsearch-6.5.1/bin/elasticsearch -Enode.name=first
 ./elasticsearch-6.5.1/bin/elasticsearch -Enode.name=second
```

Open Kibana, run this in console

```
DELETE my-index

PUT /my-index
{
  "settings": {
    "index.routing.allocation.include._name": "first",
    "index.number_of_shards": 1,
    "index.number_of_replicas": 0
  },
  "mappings": {
    "_doc": {
      "properties": {
        "field": {
          "type": "keyword"
        }
      }
    }
  }
}

# all shards assigned on first node
GET _cat/shards/my-index

# this works
POST /_scripts/painless/_execute
{
  "script": {
    "source": "doc['field'].value.length() <= params.max_length",
    "params": {
      "max_length": 4
    }
  },
  "context": "filter",
  "context_setup": {
    "index": "my-index",
    "document": {
      "field": "four"
    }
  }
}

# move index to second node
PUT /my-index/_settings
{
  "settings": {
    "index.routing.allocation.include._name": "second"
  }
}

# all shards assigned on second node
GET _cat/shards/my-index

# NPE baby!
POST /_scripts/painless/_execute
{
  "script": {
    "source": "doc['field'].value.length() <= params.max_length",
    "params": {
      "max_length": 4
    }
  },
  "context": "filter",
  "context_setup": {
    "index": "my-index",
    "document": {
      "field": "four"
    }
  }
}
```

The NPE is the following

```
[2018-11-29T14:00:49,023][WARN ][r.suppressed             ] [first] path: /_scripts/painless/_execute, params: {pretty=}
org.elasticsearch.transport.RemoteTransportException: [second][127.0.0.1:9301][cluster:admin/scripts/painless/execute[s]]
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:221) ~[?:?]
	at org.elasticsearch.index.mapper.SourceToParse.<init>(SourceToParse.java:56) ~[elasticsearch-6.5.1.jar:6.5.1]
	at org.elasticsearch.index.mapper.SourceToParse.source(SourceToParse.java:32) ~[elasticsearch-6.5.1.jar:6.5.1]
	at org.elasticsearch.painless.PainlessExecuteAction$TransportAction.prepareRamIndex(PainlessExecuteAction.java:597) ~[?:?]
	at org.elasticsearch.painless.PainlessExecuteAction$TransportAction.innerShardOperation(PainlessExecuteAction.java:547) ~[?:?]
	at org.elasticsearch.painless.PainlessExecuteAction$TransportAction.shardOperation(PainlessExecuteAction.java:536) ~[?:?]
	at org.elasticsearch.painless.PainlessExecuteAction$TransportAction.shardOperation(PainlessExecuteAction.java:475) ~[?:?]
	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$1.doRun(TransportSingleShardAction.java:112) ~[elasticsearch-6.5.1.jar:6.5.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:723) ~[elasticsearch-6.5.1.jar:6.5.1]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-6.5.1.jar:6.5.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

--- a/modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java
+++ b/modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java
@@ -218,19 +218,20 @@ public class PainlessExecuteAction extends Action<PainlessExecuteAction.Response
                 ContextSetup that = (ContextSetup) o;
                 return Objects.equals(index, that.index) &&
                     Objects.equals(document, that.document) &&
-                    Objects.equals(query, that.query);
+                    Objects.equals(query, that.query) &&
+                    Objects.equals(xContentType, that.xContentType);
             }
 
             @Override
             public int hashCode() {
-                return Objects.hash(index, document, query);
+                return Objects.hash(index, document, query, xContentType);
             }
 
             @Override
             public void writeTo(StreamOutput out) throws IOException {
                 out.writeOptionalString(index);
                 out.writeOptionalBytesReference(document);
-                out.writeOptionalString(xContentType != null ? xContentType.mediaType(): null);
+                out.writeOptionalString(xContentType != null ? xContentType.mediaTypeWithoutParameters(): null);
                 out.writeOptionalNamedWriteable(query);
             }
 
@@ -347,11 +348,13 @@ public class PainlessExecuteAction extends Action<PainlessExecuteAction.Response
         // For testing only:
         @Override
         public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
             builder.field(SCRIPT_FIELD.getPreferredName(), script);
             builder.field(CONTEXT_FIELD.getPreferredName(), context.name);
             if (contextSetup != null) {
                 builder.field(CONTEXT_SETUP_FIELD.getPreferredName(), contextSetup);
             }
+            builder.endObject();
             return builder;
         }
 

--- a/modules/lang-painless/src/test/java/org/elasticsearch/painless/PainlessExecuteRequestTests.java
+++ b/modules/lang-painless/src/test/java/org/elasticsearch/painless/PainlessExecuteRequestTests.java
@@ -20,9 +20,14 @@ package org.elasticsearch.painless;
 
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
 import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
 import org.elasticsearch.index.query.MatchAllQueryBuilder;
 import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.painless.PainlessExecuteAction.Request.ContextSetup;
@@ -30,12 +35,40 @@ import org.elasticsearch.script.Script;
 import org.elasticsearch.script.ScriptContext;
 import org.elasticsearch.script.ScriptType;
 import org.elasticsearch.search.SearchModule;
-import org.elasticsearch.test.AbstractStreamableXContentTestCase;
+import org.elasticsearch.test.AbstractStreamableTestCase;
 
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.util.Collections;
 
-public class PainlessExecuteRequestTests extends AbstractStreamableXContentTestCase<PainlessExecuteAction.Request> {
+import static org.hamcrest.Matchers.equalTo;
+
+public class PainlessExecuteRequestTests extends AbstractStreamableTestCase<PainlessExecuteAction.Request> {
+
+    // Testing XContent serialization manually here, because the xContentType field in ContextSetup determines
+    // how the request needs to parse and the xcontent serialization framework randomizes that. The XContentType
+    // is not known and accessable when the test request instance is created in the xcontent serialization framework.
+    // Changing that is a big change. Writing a custom xcontent test here is the best option for now, because as far
+    // as I know this request class is the only case where this is a problem.
+    public final void testFromXContent() throws Exception {
+        for (int i = 0; i < 20; i++) {
+            PainlessExecuteAction.Request testInstance = createTestInstance();
+            ContextSetup contextSetup = testInstance.getContextSetup();
+            XContent xContent = randomFrom(XContentType.values()).xContent();
+            if (contextSetup != null && contextSetup.getXContentType() != null) {
+                xContent = contextSetup.getXContentType().xContent();
+            }
+
+            try (XContentBuilder builder = XContentBuilder.builder(xContent)) {
+                builder.value(testInstance);
+                StreamInput instanceInput = BytesReference.bytes(builder).streamInput();
+                try (XContentParser parser = xContent.createParser(xContentRegistry(), LoggingDeprecationHandler.INSTANCE, instanceInput)) {
+                    PainlessExecuteAction.Request result = PainlessExecuteAction.Request.parse(parser);
+                    assertThat(result, equalTo(testInstance));
+                }
+            }
+        }
+    }
 
     @Override
     protected NamedWriteableRegistry getNamedWriteableRegistry() {
@@ -60,16 +93,6 @@ public class PainlessExecuteRequestTests extends AbstractStreamableXContentTestC
         return new PainlessExecuteAction.Request();
     }
 
-    @Override
-    protected PainlessExecuteAction.Request doParseInstance(XContentParser parser) throws IOException {
-        return PainlessExecuteAction.Request.parse(parser);
-    }
-
-    @Override
-    protected boolean supportsUnknownFields() {
-        return false;
-    }
-
     public void testValidate() {
         Script script = new Script(ScriptType.STORED, null, randomAlphaOfLength(10), Collections.emptyMap());
         PainlessExecuteAction.Request request = new PainlessExecuteAction.Request(script, null, null);
@@ -78,20 +101,24 @@ public class PainlessExecuteRequestTests extends AbstractStreamableXContentTestC
         assertEquals("Validation Failed: 1: only inline scripts are supported;", e.getMessage());
     }
 
-    private static ContextSetup randomContextSetup() {
+    private static ContextSetup randomContextSetup()  {
         String index = randomBoolean() ? randomAlphaOfLength(4) : null;
         QueryBuilder query = randomBoolean() ? new MatchAllQueryBuilder() : null;
-        // TODO: pass down XContextType to createTestInstance() method.
-        // otherwise the document itself is different causing test failures.
-        // This should be done in a separate change as the test instance is created before xcontent type is randomly picked and
-        // all the createTestInstance() methods need to be changed, which will make this a big chnage
-//        BytesReference doc = randomBoolean() ? new BytesArray("{}") : null;
         BytesReference doc = null;
+        XContentType xContentType = randomFrom(XContentType.values());
+        if (randomBoolean()) {
+            try {
+                XContentBuilder xContentBuilder = XContentBuilder.builder(xContentType.xContent());
+                xContentBuilder.startObject();
+                xContentBuilder.endObject();
+                doc = BytesReference.bytes(xContentBuilder);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
 
         ContextSetup contextSetup = new ContextSetup(index, doc, query);
-//        if (doc != null) {
-//            contextSetup.setXContentType(XContentType.JSON);
-//        }
+        contextSetup.setXContentType(xContentType);
         return contextSetup;
     }
 }

